### PR TITLE
[SPARK-20762][ML]Make String Params Case-Insensitive

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
Make String Params Case-Insensitive:
`solver`, `modelType`, `initMode`, `metricName`, `handleInvalid`, `strategy`, `stringOrderType`, `coldStartStrategy`, `impurity`, `lossType`, `featureSubsetStrategy`, `intermediateStorageLevel`, `finalStorageLevel`

Leave alone `ChiSqSelector` & `StringIndexer`, for they are not easy to handle like others.

## How was this patch tested?
existing tests and added tests

### CODE REVIEW

**Comment by zhengruifeng** [comment]
Jenkins, retest this please

**Comment by zhengruifeng** [comment]
ping @yanboliang

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/MulticlassClassificationEvaluator.scala [hunk 1]
```diff
@@ -44,12 +46,10 @@ class MulticlassClassificationEvaluator @Since("1.5.0") (@Since("1.5.0") overrid
...
-  val metricName: Param[String] = {
-    val allowedParams = ParamValidators.inArray(Array("f1", "weightedPrecision",
-      "weightedRecall", "accuracy"))
-    new Param(this, "metricName", "metric name in evaluation " +
-      "(f1|weightedPrecision|weightedRecall|accuracy)", allowedParams)
-  }
+  val metricName: Param[String] = new Param[String](this, "metricName", "metric name in" +
+    " evaluation (f1|weightedPrecision|weightedRecall|accuracy)",
+    (value: String) => Array("f1", "weightedprecision", "weightedrecall", "accuracy")
+      .contains(value.toLowerCase(Locale.ROOT)))
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/MulticlassClassificationEvaluator.scala]: Ditto.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala [hunk 2]
```diff
@@ -48,10 +50,11 @@ final class RegressionEvaluator @Since("1.4.0") (@Since("1.4.0") override val ui
    * @group param
    */
   @Since("1.4.0")
-  val metricName: Param[String] = {
-    val allowedParams = ParamValidators.inArray(Array("mse", "rmse", "r2", "mae"))
-    new Param(this, "metricName", "metric name in evaluation (mse|rmse|r2|mae)", allowedParams)
-  }
+  val metricName: Param[String] = new Param[String](this, "metricName", "metric name in" +
+    " evaluation (mse|rmse|r2|mae)",
+    (value: String) => Array("mse", "rmse", "r2", "mae")
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala]: Ditto.

File: mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala [hunk 3]
```diff
@@ -128,7 +130,8 @@ private[feature] trait ChiSqSelectorParams extends Params
   final val selectorType = new Param[String](this, "selectorType",
     "The selector type of the ChisqSelector. " +
       "Supported options: " + OldChiSqSelector.supportedSelectorTypes.mkString(", "),
-    ParamValidators.inArray[String](OldChiSqSelector.supportedSelectorTypes))
+    (value: String) => OldChiSqSelector.supportedSelectorTypes.map(_.toLowerCase(Locale.ROOT))
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/feature/ChiSqSelector.scala]: Supported selector types should always be stored as lower case, please update corresponding code snippet in ```mllib.feature.ChiSqSelector``` from:
```
private[spark] val NumTopFeatures: String = "numTopFeatures"
......
```
to
```
private[spark] val NumTopFeatures: String = "numTopFeatures".toLowerCase
......
```

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/BinaryClassificationEvaluator.scala [hunk 4]
```diff
@@ -70,6 +71,10 @@ class BinaryClassificationEvaluator @Since("1.4.0") (@Since("1.4.0") override va
 
   setDefault(metricName -> "areaUnderROC")
 
+  private def getFormattedMetricName =
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/BinaryClassificationEvaluator.scala]: Is this really necessary?

File: mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala [hunk 5]
```diff
@@ -45,7 +47,7 @@ private[feature] trait ImputerParams extends Params with HasInputCols {
   final val strategy: Param[String] = new Param(this, "strategy", s"strategy for imputation. " +
     s"If ${Imputer.mean}, then replace missing values using the mean value of the feature. " +
     s"If ${Imputer.median}, then replace missing values using the median value of the feature.",
-    ParamValidators.inArray[String](Array(Imputer.mean, Imputer.median)))
+    (value: String) => Array(Imputer.mean, Imputer.median).contains(value.toLowerCase(Locale.ROOT)))
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala]: Ditto.

File: mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala [hunk 6]
```diff
@@ -313,7 +313,11 @@ class GeneralizedLinearRegression @Since("2.0.0") (@Since("2.0.0") override val
    * @group setParam
    */
   @Since("2.0.0")
-  def setSolver(value: String): this.type = set(solver, value)
+  def setSolver(value: String): this.type = {
+    require("irls" == value.toLowerCase(Locale.ROOT),
+      s"Solver $value was not supported. Supported options: irls")
+    set(solver, value)
+  }
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala]: Actually we can't do this, since MLlib supports set params via other entrances. Currently we can leave as it is, until we resolved #16028.

File: mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala [hunk 7]
```diff
@@ -81,7 +83,8 @@ private[classification] trait MultilayerPerceptronParams extends PredictorParams
   final val solver: Param[String] = new Param[String](this, "solver",
     "The solver algorithm for optimization. Supported options: " +
       s"${MultilayerPerceptronClassifier.supportedSolvers.mkString(", ")}. (Default l-bfgs)",
-    ParamValidators.inArray[String](MultilayerPerceptronClassifier.supportedSolvers))
+    (value: String) => MultilayerPerceptronClassifier.supportedSolvers
+      .contains(value.toLowerCase(Locale.ROOT)))
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: What do you think of adding a new function in ```object ParamValidators``` as
```
def inStringArray(allowed: Array[String]): String => Boolean = { (value: String) =>
    allowed.contains(value.toLowerCase(java.util.Locale.ROOT))
  }
```
to facilitate similar check here and other place.

Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala]: I think it a good idea.

**Comment by zhengruifeng** [comment]
@yanboliang I update this PR and revert changes on `setSolver` in GLR and LiR. Thanks for your reviewing.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 8]
```diff
@@ -93,8 +93,8 @@ private[classification] trait LogisticRegressionParams extends ProbabilisticClas
   @Since("2.1.0")
   final val family: Param[String] = new Param(this, "family",
     "The name of family which is a description of the label distribution to be used in the " +
-      s"model. Supported options: ${supportedFamilyNames.mkString(", ")}.",
-    (value: String) => supportedFamilyNames.contains(value.toLowerCase(Locale.ROOT)))
+      s"model. Supported options: ${supportedFamilyNames.mkString("(", ",", ")")}.",
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Please keep the original output format for supported family names.

File: mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala [hunk 9]
```diff
@@ -199,7 +199,7 @@ private[regression] trait GeneralizedLinearRegressionBase extends PredictorParam
         require(supportedFamilyAndLinkPairs.contains(
           Family.fromParams(this) -> Link.fromParams(this)),
           s"Generalized Linear Regression with ${$(family)} family " +
-            s"does not support ${$(link)} link function.")
+            s"does not support ${getLink.toLowerCase(Locale.ROOT)} link function.")
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala]: Please revert this change.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 10]
```diff
@@ -460,7 +463,7 @@ object LinearRegression extends DefaultParamsReadable[LinearRegression] {
   val MAX_FEATURES_FOR_NORMAL_SOLVER: Int = WeightedLeastSquares.MAX_NUM_FEATURES
 
   /** String name for "auto". */
-  private[regression] val AUTO = "auto"
+  private[regression] val Auto = "auto"
 
   /** String name for "normal". */
   private[regression] val NORMAL = "normal"
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: ```NORMAL``` -> ```Normal```
